airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/file_based_cursor.py

```diff
@@ -0,0 +1,75 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from datetime import datetime, timedelta
+from typing import Any, List, Mapping
+
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+
+
+class FileBasedCursor:
+    def __init__(self, max_history_size: int, time_window_if_history_is_full: timedelta, logger: logging.Logger):
+        self._file_to_datetime_history: Mapping[str:datetime] = {}
+        self._max_history_size = max_history_size
+        self._time_window_if_history_is_full = time_window_if_history_is_full
+        self._history_is_partial = False
+        self._logger = logger
+        self._start_time = self._compute_start_time()
+
+    def set_initial_state(self, value: Mapping[str, Any]):
+        self._file_to_datetime_history = value.get("history", {})
+        self._history_is_partial = value.get("history_is_partial", False)
+        self._start_time = self._compute_start_time()
+
+    def add_file(self, file: RemoteFile):
+        self._file_to_datetime_history[file.uri] = file.last_modified.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        if len(self._file_to_datetime_history) > self._max_history_size:
+            oldest_file = min(self._file_to_datetime_history, key=self._file_to_datetime_history.get)
+            del self._file_to_datetime_history[oldest_file]
+
+    def to_dict(self):
+        state = {
+            "history": self._file_to_datetime_history,
+            "history_is_partial": self._history_is_partial,
+        }
+        return state
+
+    def is_history_partial(self):
+        return self._history_is_partial
+
+    def get_files_to_sync(self, all_files: List[RemoteFile]):
+        start_time = self._compute_start_time()
+        files_to_sync = [
+            f
+            for f in all_files
+            if (
+                f.last_modified >= start_time
```

Comment:
opening the door to challenging this implementation:

According to the AbstractFileBasedStreamReader.list_matching_files's interface, this all the files passed to this method should have a timestamp >= start_time. 

However, it seems reasonable for the cursor to do this filtering since it owns the start_time.

I also thought about removing the start_time parameter from `AbstractFileBasedStreamReader.list_matching_files`, but I can imagine it being useful for sources that allow filtering based on a timestamp (I don't know if there are any)